File: 백준 문제 풀이 모음/kotlin/class 4/1238.kt
import java.util.PriorityQueue

private lateinit var result_dist : MutableList<MutableList<Int>>
private lateinit var list : MutableList<MutableList<Node>>
private val INF = Integer.MAX_VALUE

private class Node (var nodeNum: Int, var weight : Int) : Comparable<Node> {
    override fun compareTo(other: Node): Int {
        return weight - other.weight
    }

}
fun main() = with(System.`in`.bufferedReader()){
    val input = readLine().split(' ').toList()
    val n = input[0].toInt() // n명의 학생
    val m = input[1].toInt() // m edge
    val x = input[2].toInt() // end

    list = MutableList<MutableList<Node>>(n+1) { MutableList<Node> (0) { Node(0,0)} }
    result_dist = MutableList<MutableList<Int>>(n+1) { MutableList<Int>(n+1) { INF } }


    repeat(m) {
        val input = readLine().split(' ').toList()
        list[input[0].toInt()].add(Node(input[1].toInt(), input[2].toInt()))
    }

    for(i in 1 .. n) {
        val eachdist = dijkstra(i, n)
        for (j in 1 .. n) {
            result_dist[i][j] = eachdist[j]
        }
    }

//    result_dist.forEach {
//        println(it)
//    }
    var result = -1
    for (i in 1 .. n) {
        result = Math.max(result, result_dist[x][i] + result_dist[i][x])
    }

    println(result)

}

fun dijkstra(start: Int, n:Int) : MutableList<Int> {
    val dist = MutableList<Int>(n+1) { INF }
    val visited = MutableList<Boolean>(n+1) { false }

    dist[start] = 0
    val pq = PriorityQueue<Node>()
    pq.offer(Node(start, 0))

    while (pq.isNotEmpty()) {
        val node = pq.poll()
        if (!visited[node.nodeNum]) {
            visited[node.nodeNum] = true

            list[node.nodeNum].forEach {
                if (!visited[it.nodeNum] && dist[it.nodeNum] > dist[node.nodeNum] + it.weight) {
                    dist[it.nodeNum] = dist[node.nodeNum] + it.weight
                    pq.add(Node(it.nodeNum, dist[it.nodeNum]))
                }
            }

        }
    }

    return dist
}

File: 백준 문제 풀이 모음/python/class 4/1238.py
import sys, heapq
INF = sys.maxsize
input = sys.stdin.readline

def dijkstra(start_node) :
    distance =[INF] * (N+1) 
    distance[start_node] = 0
    heap = []

    while heap :
        now_cost, node = heapq.heappop(heap)

        for next_cost, next_node in graph[node] :
            next_cost = next_cost + now_cost
            if next_cost < distance[next_node] :
                distance[next_node] = next_cost
                heapq.heappush(heap, [next_cost, next_node])
        return distance




N, M, X = map(int, input().split())

graph = [[] * (N+1) for _ in range(N+1)]


for i in range(M):
    start, end, cost = map(int, input().split())
    graph[start].append([cost,end])

answer = [0] * (N+1)

for i in range(1, N+1) :
    arr = dijkstra(i)
    answer[i] += arr[X]
    arr2 = dijkstra(X)
    answer[i] += arr[i]

print(max(answer))